### PR TITLE
Parquet: Move to ValueReader generation to a visitor

### PR DESCRIPTION
**Pull request by Fokko**
This will replace the switch statement that relies on the deprecated `getOriginalType()` with a visitor pattern of the new `LogicalTypeAnnotation`. OriginalType does not include support for Nanosecond precision timestamps.

### CODE REVIEW

File: parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java [hunk 1]
```diff
@@ -108,6 +110,110 @@ public ParquetValueReader<?> struct(
...
+        ColumnDescriptor desc,
+        org.apache.iceberg.types.Type.PrimitiveType expected,
+        PrimitiveType primitive) {
+      this.desc = desc;
+      this.expected = expected;
+      this.primitive = primitive;
+    }
+
+    @Override
+    public Optional<ParquetValueReader<?>> visit(
```
Comment (nk1506) [inline parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java:130]: is this `case UTF8:
            return new ParquetValueReaders.StringReader(desc);` being covered here ?

Comment (Fokko) [reply parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java:130]: Yes, the old `UTF8` `OriginalType`, maps to a `StringType`: https://github.com/apache/parquet-mr/blob/65bc51846010360f3dd4304103ec3c637776d7c9/parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java#L197-L198

**Review by nk1506 — APPROVED** [review]

**Comment by Fokko** [comment]
@aokolnychyi do you have a spare cycle for reviewing this?

**Review by jacobmarble — COMMENTED** [review]
Looks good to me!

File: parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java [hunk 2]
```diff
@@ -108,6 +110,113 @@ public ParquetValueReader<?> struct(
...
+        LogicalTypeAnnotation.TimestampLogicalTypeAnnotation timestampLogicalType) {
+      if (timestampLogicalType.getUnit() == LogicalTypeAnnotation.TimeUnit.MICROS) {
+        Types.TimestampType tsMicrosType = (Types.TimestampType) expected;
+        return tsMicrosType.shouldAdjustToUTC()
+            ? Optional.of(new TimestamptzReader(desc))
+            : Optional.of(new TimestampReader(desc));
+      } else if (timestampLogicalType.getUnit() == LogicalTypeAnnotation.TimeUnit.MICROS) {
```
Comment (amogh-jahagirdar) [inline parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java:186]: I think this should be == LogicalTypeAnnotation.TimeUnit.MILLIS?

Comment (amogh-jahagirdar) [inline parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java:186]: I think this should be == LogicalTypeAnnotation.TimeUnit.MILLIS? Not Micros

**Review by amogh-jahagirdar — COMMENTED** [review]
I think there's just a bug in the visit(timestampLogicalType) case, but everything else looks great to me.  This does lead me to wonder though, do the parquet reader tests test all these data types? 

I'll double check that. We don't need to need to couple any of those in this PR though.

File: parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java (hunk 2 above)
Comment (Fokko) [reply parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java]: Oof, that's a good one! I would this part to be thoroughly tested as well.

**Review by amogh-jahagirdar — APPROVED** [review]

**Comment by Fokko** [comment]
Thanks @jacobmarble, @nk1506 & @amogh-jahagirdar for the review 🙌

**Pull request merged**